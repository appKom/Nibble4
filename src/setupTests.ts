// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import { CartItem, createCartItem } from "types/inventory";
import { modalTypes } from "types/modal";
import { Product } from "types/inventory";
import { User } from "types/user";

const mockProduct: Product = {
  pk: 95,
  name: "Powerking",
  description: "Energidrikk",
  price: 10,
  image: {
    sm: "/media/images/responsive/sm/38688f77-21b6-4936-8ba6-c976538fb110.png",
  },
  category: {
    pk: 3,
    name: "Drikke",
  },
};

const mockUser: User = {
  pk: 1337,
  balance: 1337,
  first_name: "SuperAdmin",
};

const mockCartItem: CartItem = createCartItem(95);

const mockStateWithUser: State = {
  user: mockUser,
  inventory: [],
  category: "Alt",
  cart: {},
  modalState: modalTypes.DISABLED,
};

const mockState: State = {
  user: mockUser,
  inventory: [mockProduct],
  category: "Alt",
  cart: { 95: mockCartItem },
  modalState: modalTypes.DISABLED,
};

const emptyState: State = {
  user: null,
  inventory: [],
  category: "Alt",
  cart: {},
  modalState: modalTypes.DISABLED,
};

export const userContext: StateContextProps = {
  state: mockStateWithUser,
  dispatch: () => null,
};

export const fullContext: StateContextProps = {
  state: mockState,
  dispatch: () => null,
};

export const emptyContext: StateContextProps = {
  state: emptyState,
  dispatch: () => null,
};
