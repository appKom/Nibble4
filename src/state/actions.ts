import { Product } from "types/inventory";
import { User } from "types/user";
import { modalTypes } from "types/modal";

export enum ActionTypes {
  SET_USER = "SET_USER",
  LOGOUT_USER = "LOGOUT_USER",
  SET_INVENTORY = "SET_INVENTORY",
  SET_FAVOURITES = "SET_FAVOURITES",
  ADD_TO_CART = "ADD_TO_CART",
  REMOVE_FROM_CART = "REMOVE_FROM_CART",
  EMPTY_CART = "EMPTY_CART",
  SET_CATEGORY = "SET_CATEGORY",
  SET_MODAL_STATE = "SET_MODAL_STATE",
}

type SetUserAction = { type: ActionTypes.SET_USER; payload: User };
type LogoutUserAction = { type: ActionTypes.LOGOUT_USER };
type SetInventoryAction = {
  type: ActionTypes.SET_INVENTORY;
  payload: Product[];
};

type SetFavouritesAction = {
  type: ActionTypes.SET_FAVOURITES;
  payload: Product[];
};

type SetCategoryAction = {
  type: ActionTypes.SET_CATEGORY;
  payload: string;
};
type AddToCartAction = {
  type: ActionTypes.ADD_TO_CART;
  payload: number;
};
type RemoveFromCartAction = {
  type: ActionTypes.REMOVE_FROM_CART;
  payload: number;
};

type EmptyCartAction = {
  type: ActionTypes.EMPTY_CART;
};

type SetModalState = {
  type: ActionTypes.SET_MODAL_STATE;
  payload: modalTypes;
};

export type Action =
  | SetUserAction
  | LogoutUserAction
  | SetInventoryAction
  | SetFavouritesAction
  | SetCategoryAction
  | AddToCartAction
  | RemoveFromCartAction
  | EmptyCartAction
  | SetModalState;

export const setUser = (user: User): SetUserAction => ({
  type: ActionTypes.SET_USER,
  payload: user,
});
export const logoutUser = (): LogoutUserAction => ({
  type: ActionTypes.LOGOUT_USER,
});
export const setInventory = (inventory: Product[]): SetInventoryAction => ({
  type: ActionTypes.SET_INVENTORY,
  payload: inventory,
});

export const setFavourites = (favourites: Product[]): SetFavouritesAction => ({
  type: ActionTypes.SET_FAVOURITES,
  payload: favourites,
});

export const addToCart = (id: number): AddToCartAction => ({
  type: ActionTypes.ADD_TO_CART,
  payload: id,
});
export const removeFromCart = (id: number): RemoveFromCartAction => ({
  type: ActionTypes.REMOVE_FROM_CART,
  payload: id,
});
export const emptyCart = (): EmptyCartAction => ({
  type: ActionTypes.EMPTY_CART,
});

export const setCategory = (id: string): SetCategoryAction => ({
  type: ActionTypes.SET_CATEGORY,
  payload: id,
});

export const setModalState = (modalType: modalTypes): SetModalState => ({
  type: ActionTypes.SET_MODAL_STATE,
  payload: modalType,
});
