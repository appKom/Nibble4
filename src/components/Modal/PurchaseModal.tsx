import React, { useContext, FC } from "react";
import styled from "styled-components";
import { StateContext } from "state/state";
import { calculateCartTotal } from "types/inventory";
import { OffWhite } from "utility/style";
import Button from "components/Button";
import purchaseItems from "api/order";
import { setModalState } from "state/actions";
import { modalTypes } from "types/modal";
import useSound from "use-sound";

const styledButton = {
  gridColumn: 2,
  gridRow: 3,
  width: "80%",
  height: "85%",
  outline: "none",
  marginLeft: "auto",
  marginRight: "auto",
  boxShadow: "2px 2px 7px #888888",
};

const styledText = {
  gridColumn: 2,
  gridRow: 1,
};

const PurchaseModal: FC = () => {
  const { state, dispatch } = useContext(StateContext);
  const { cart, inventory, user } = state;

  const totalPrice = calculateCartTotal(cart, inventory);

  const [playKing] = useSound(
    `${process.env.PUBLIC_URL}/audio/purchase_king.mp3`,
    {
      volume: 1,
    }
  );

  const [playNice] = useSound(
    `${process.env.PUBLIC_URL}/audio/purchase_nice.mp3`,
    {
      volume: 1,
    }
  );

  const purchase = () => {
    purchaseItems(user!.pk, cart)
      .then((response) => {
        if (response.ok) {
          totalPrice >= 100 ? playKing() : playNice();
          dispatch(setModalState(modalTypes.COMPLETE));
        } else dispatch(setModalState(modalTypes.ERROR));
      })
      .catch(() => dispatch(setModalState(modalTypes.ERROR)));
  };

  return (
    <Container>
      <h3 style={styledText}> Bekreft kjøp </h3>
      <TotalDiv>
        <span> Total: </span>
        <hr />
        <span> {totalPrice} kr </span>
      </TotalDiv>
      <Button style={styledButton} onClick={purchase}>
        Kjøp
      </Button>
    </Container>
  );
};

export default PurchaseModal;

const Container = styled.div`
  text-align: center;

  display: grid;
  grid-template-rows: 1fr 3fr 1fr;
  grid-template-columns: 1fr 3fr 1fr;
`;

const TotalDiv = styled.div`
  display: grid;

  width: 100%;
  height: 60%;

  background-color: ${OffWhite};
  margin-left: auto;
  margin-right: auto;
  border-radius: 2px;

  grid-column: 2;
  grid-row: 2;
  hr {
    width: 100%;
    overflow: hidden;
    border-top: 1px dashed #aeb2a8;
    height: 0px;
  }
  box-shadow: 2px 2px 7px #888888;
`;
