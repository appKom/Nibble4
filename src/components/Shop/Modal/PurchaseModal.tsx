import React, { useContext, FC } from "react";
import styled from "styled-components";
import { StateContext } from "state/state";
import { calculateCartTotal } from "types/inventory";
import { OffWhite } from "utility/style";
import Button from "../../../atoms/Button";
import purchaseItems from "api/order";
import { setPurchaseComplete, setModalState } from "state/actions";
import { modalTypes } from "types/modal";

const styledButton = {
  gridColumn: 2,
  gridRow: 3,
  width: "80%",
  height: "85%",
  outline: "none",
  marginLeft: "auto",
  marginRight: "auto",
  boxShadow: "2px 2px 7px #888888",
};

const styledText = {
  gridColumn: 2,
  gridRow: 1,
};

const PurchaseModal: FC = () => {
  const { state, dispatch } = useContext(StateContext);
  const { cart, inventory, user } = state;

  const showPurchaseComplete = (show: boolean) =>
    dispatch(setPurchaseComplete(show));

  const totalPrice = calculateCartTotal(cart, inventory);
  const purchase = () => {
    purchaseItems(user!.pk, cart).then((response) => {
      if (response.status == 201) dispatch(setModalState(modalTypes.COMPLETE));
    });
  };

  return (
    <Container>
      <h3 style={styledText}> Confirm purchase </h3>
      <TotalDiv>
        <span> Your total </span>
        <hr />
        <span> {totalPrice} kr </span>
      </TotalDiv>
      <Button style={styledButton} onClick={purchase}>
        Purchase
      </Button>
    </Container>
  );
};

export default PurchaseModal;

const Container = styled.div`
  text-align: center;

  display: grid;
  grid-template-rows: 1fr 3fr 1fr;
  grid-template-columns: 1fr 3fr 1fr;
`;

const TotalDiv = styled.div`
  display: grid;

  width: 100%;
  height: 60%;

  background-color: ${OffWhite};
  margin-left: auto;
  margin-right: auto;
  border-radius: 2px;

  grid-column: 2;
  grid-row: 2;
  hr {
    width: 100%;
    overflow: hidden;
    border-top: 1px dashed #aeb2a8;
    height: 0px;
  }
  box-shadow: 2px 2px 7px #888888;
`;
