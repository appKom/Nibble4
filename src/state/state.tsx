import React, { createContext, Dispatch, ReactNode, useReducer } from "react";
import {
  Product,
  CartItem,
  incrementCartItem,
  createCartItem,
  decrementCartItem,
} from "types/inventory";
import { User, OlCoinsUser } from "types/user";
import { modalTypes } from "types/modal";
import { Action, ActionTypes } from "state/actions";

type State = {
  user?: User;
  olCoinsUser?: OlCoinsUser;
  inventory: Product[];
  category: string;
  cart: { [id: number]: CartItem };
  modalState: modalTypes;
  newOlCoins: number;
};

const initialState: State = {
  olCoinsUser: { balance: 100 },
  inventory: [],
  category: "Alt",
  cart: {},
  modalState: modalTypes.DISABLED,
  newOlCoins: 0,
};

const getIncrementedCartItem = (id: number, state: State): CartItem => {
  if (state.cart[id]) {
    return incrementCartItem(state.cart[id]);
  }
  return createCartItem(id);
};

const getDecrementedCart = (id: number, state: State) => {
  if (state.cart[id].quantity > 1) {
    return { ...state.cart, [id]: decrementCartItem(state.cart[id]) };
  }
  const removedCart = { ...state.cart };
  delete removedCart[id];
  return { ...removedCart };
};

const updateNewOlCoins = (newValue: number, state: State) => {
  return state.newOlCoins + newValue;
};

const StateReducer = (state: State, action: Action): State => {
  switch (action.type) {
    case ActionTypes.SET_USER:
      return { ...state, user: action.payload };
    case ActionTypes.LOGOUT_USER:
      return {
        ...state,
        user: undefined,
        cart: {},
        category: "Alt",
        modalState: modalTypes.DISABLED,
      };
    case ActionTypes.SET_INVENTORY:
      return { ...state, inventory: action.payload };
    case ActionTypes.SET_CATEGORY:
      return { ...state, category: action.payload };
    case ActionTypes.SET_MODAL_STATE:
      return { ...state, modalState: action.payload };
    case ActionTypes.CHANGE_NEWOLCOINS:
      return {
        ...state,
        newOlCoins: updateNewOlCoins(action.payload, state),
      };
    case ActionTypes.ADD_TO_CART:
      return {
        ...state,
        cart: {
          ...state.cart,
          [action.payload]: getIncrementedCartItem(action.payload, state),
        },
      };
    case ActionTypes.REMOVE_FROM_CART: {
      return {
        ...state,
        cart: getDecrementedCart(action.payload, state),
      };
    }
    case ActionTypes.EMPTY_CART:
      return { ...state, cart: {} };
    default:
      return { ...state };
  }
};

type StateContextProps = {
  state: State;
  dispatch: Dispatch<Action>;
};

const initialContext: StateContextProps = {
  state: initialState,
  dispatch: () => null,
};

export const StateContext = createContext<StateContextProps>(initialContext);

type StateProviderProps = {
  children: ReactNode;
};

export const StateProvider = (props: StateProviderProps): JSX.Element => {
  const [state, dispatch] = useReducer(StateReducer, initialState);
  const { children } = props;

  return (
    <StateContext.Provider value={{ state, dispatch }}>
      {children}
    </StateContext.Provider>
  );
};
