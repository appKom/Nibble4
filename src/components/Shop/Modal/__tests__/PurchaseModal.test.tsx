import React from "react";
import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import PurchaseModal from "../PurchaseModal";
import { StateContext } from "state/state";
import { fullContext } from "setupTests";

afterEach(cleanup);

it("Should match snapshot", () => {
  const { asFragment } = render(
    <StateContext.Provider value={fullContext}>
      <PurchaseModal />
    </StateContext.Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});
