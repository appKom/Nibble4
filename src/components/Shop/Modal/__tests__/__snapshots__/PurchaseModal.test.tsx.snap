// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-gsTCUz exa-Dkc"
  >
    <h3
      style="grid-column: 2; grid-row: 1;"
    >
       Bekreft kjøp 
    </h3>
    <div
      class="sc-dlfnbm eWUpqm"
    >
      <span>
         Total: 
      </span>
      <hr />
      <span>
         10 kr 
      </span>
    </div>
    <button
      class="sc-bdfBwQ eAPZdX"
      style="grid-column: 2; grid-row: 3; width: 80%; height: 85%; outline: none; margin-left: auto; margin-right: auto; box-shadow: 2px 2px 7px #888888;"
    >
      Kjøp
    </button>
  </div>
</DocumentFragment>
`;
